### PR TITLE
add the "dar" command for doing development with cyrus-docker image

### PR DESCRIPTION
**Pull request by rjbs**
The main thrust of this branch is to add **dar**, a program for managing a cyrus-docker-based build and test.  The `bin/dar` program has a more detailed description, but in brief:

* put `dar` in your path
* in your repo, run `dar --start`
* later, run `dar smoke` or `dar sh` or various other commands
* when you're all done, you can run `dar --prune`

To help with this, `cyd` gets some changes:

* checkout is now called clone (because it will never change branch)
* `cyr idle` is added as the default startup command for dar-helping containers
* `git clean` before builds, because too many things rely on it
* give up early if `git-version.sh` fails, because it leads to bogus test behavior

### CODE REVIEW

**Comment by rjbs** [comment]
Still to do:

- [ ] bring back `git clean` before build by default, because too many things need it :(
- [ ] run `cyr_info version` after a successful build
- [ ] abort builds early if `./tools/git-version.sh` reports "unknown"
- [ ] refuse to test if the version under test is "unknown" (maybe provide an override)

**Comment by rjbs** [comment]
I realized I should fix a few more things before getting review…

File: bin/dar [hunk 1]
```diff
@@ -5,6 +5,30 @@ package Cyrus::Docker::dar;
 
 use utf8;
 
+BEGIN {
```
Comment (wolfsage) [inline bin/dar:8]: a cpanfile would be nice too but this works

**Review by wolfsage — COMMENTED** [review]
Overall, pretty cool. I think it's a little weird to adjust to for me, but I can make the adjustment.

I really want `dar exec ls` to run *ls* on the remote system, not `cyd ls`.

Maybe `dar rexec` (for raw exec?) would be useful to have.

(ls is just an example...)

Also maybe a config option for default path of `dar` when it drops you into a shell (or runs commands). 99% of the time I want to land in cyrus-imapd, not /srv/ for example...

Pretty cool!

**Comment by rjbs** [comment]
- [ ] add something to run commands without cyd

File: bin/dar [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+  );
+
+  chomp $container_id;
+  say "✅ Container started, id: $container_id";
+
+  Process::Status->assert_ok("❌ Starting idle container");
+
+  my $container = $class->_existing_container;
+
+  unless ($container) {
```
[the remark this answers is not in the document]
Comment (rjbs) [reply bin/dar:142]: (If you didn't pass --image, it should've used the container from GHCR, or something you had tagged `ghcr.io/cyrusimap/cyrus-docker:nightly`.)

Anyway: the idea of using another log driver might be a good idea, but for now I think I'm not ready to really think about it.  If I did something, I'd add `--debug` and have it _not_ detach the idle process, maybe.

File: bin/dar (hunk 1 above)
Comment (rjbs) [reply bin/dar:8]: I feel like there can/will be a better install/setup story later :)

**Comment by rjbs** [comment]
Rather than add an option for cwd, I just set it to /srv/cyrus-imapd by default.

File: bin/dar [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+  }
+
+  if ($container->{Command} ne q{"cyd idle"}) {
+    # I don't think this should ever happen either...
+    die "❌ Weird: existing container isn't running cyd idle.  Giving up.\n";
+  }
+
+  say "⏳ Executing command in container $container->{ID}...";
+
+  my @to_run = @$args && $args->[0] eq 'run'  ? @$args
```
Comment (wolfsage) [inline bin/dar:233]: This passes `run` down to the container which doesn't exist I think?

Maybe you want
```
-  my @to_run = @$args && $args->[0] eq 'run'  ? @$args
+  my @to_run = @$args && $args->[0] eq 'run'  ? @$args[1..$#$args]
```

File: bin/dar [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+only manage one container per checkout (as identified by absolute path).
+
+The simplest thing to do is to run "dar smoke", which will start a container
+(if needed), then configure, build, and test Cyrus using the "cyd smoke"
+command packed inside the cyrus-docker image.  If you want to pass arguments to
+smoke, you can run "dar smoke -- test1 test2 test3", for example.
+
+You can run other cyd commands through dar.  If you want to get a shell and
+look into what went wrong on your last test run, you can run "dar shell".  If
+you want to run something other than a cyd command, use "dar run XYZ ABC" to
```
Comment (wolfsage) [inline bin/dar:57]: Need `dar run -- foo --bar` if passing in arguments, might be worth noting

File: bin/dar (hunk 3 above)
Comment (rjbs) [reply bin/dar:233]: Yes, thanks for the catch.

**Pull request merged**